Mathlib/Geometry/Manifold/ChartedSpace.lean

```diff
@@ -1271,4 +1322,57 @@ def Structomorph.trans (e : Structomorph G M M') (e' : Structomorph G M' M'') :
+-- xxx: better name? how does this differ from `restriction_in_maximalAtlas`?
+theorem StructureGroupoid.restriction_chart {e : PartialHomeomorph M H} (he : e ∈ atlas H M)
+    (hs : Set.Nonempty e.source) [HasGroupoid M G] [ClosedUnderRestriction G] :
+    let s := { carrier := e.source, is_open' := e.open_source : Opens M };
+    let t := { carrier := e.target, is_open' := e.open_target  : Opens H };
+    ∀ c' ∈ atlas H t,
+      (e.toHomeomorphSourceTarget).toPartialHomeomorph ≫ₕ c' ∈ G.maximalAtlas s := by
+  intro s t c' hc'
+  have : Nonempty s := nonempty_coe_sort.mpr hs
+  have : Nonempty t := nonempty_coe_sort.mpr (e.mapsTo.nonempty hs)
+  -- Choose `x ∈ t` so `c'` is the restriction of `chartAt H x`.
+  obtain ⟨x, hc'⟩ := Opens.chart_eq' hc'
+  -- As H has only one chart, `chartAt H x` is the identity: i.e., `c'` is the inclusion.
+  rw [hc', (chartAt_self_eq)]
+  -- Argue that our expression equals this chart above, at least on its source.
+  rw [PartialHomeomorph.subtypeRestr_def, PartialHomeomorph.trans_refl]
+  let goal := (e.toHomeomorphSourceTarget.toPartialHomeomorph ≫ₕ t.partialHomeomorphSubtypeCoe)
+  have : goal ≈ e.subtypeRestr s :=
+    (goal.eqOnSource_iff (e.subtypeRestr s)).mpr ⟨by simp, by intro _ _; rfl⟩
+  exact G.mem_maximalAtlas_of_eqOnSource (M := s) this (G.restriction_in_maximalAtlas he)
+
+/-- Each chart of a charted space is a structomorphism between its source and target. -/
+def PartialHomeomorph.toStructomorph {e : PartialHomeomorph M H} (he : e ∈ atlas H M)
+    [HasGroupoid M G] [ClosedUnderRestriction G] :
+    let s : Opens M := { carrier := e.source, is_open' := e.open_source }
+    let t : Opens H := { carrier := e.target, is_open' := e.open_target }
+    Structomorph G s t := by
+  intro s t
+  by_cases h : s = (∅ : Set M)
+  · have : IsEmpty s := isEmpty_coe_sort.mpr h
+    have : t = (∅ : Set H) := by
+      show e.target = ∅
+      exact e.image_source_eq_target ▸ (image_eq_empty.mpr h)
+    have : IsEmpty t := isEmpty_coe_sort.mpr this
+    exact {
+      Homeomorph.empty with
+      -- `c'` cannot exist: it would be the restriction of `chartAt H x` at some `x ∈ t`.
+      mem_groupoid := fun _ c' _ ⟨_, ⟨x, _⟩, _⟩ ↦ (this.false x).elim
+    }
+  · exact {
+      e.toHomeomorphSourceTarget with
+      mem_groupoid := by
+        intro c c' hc hc'
+        show (c.symm).trans (e.toHomeomorphSourceTarget.toPartialHomeomorph.trans c') ∈ G
+        -- The atlas of H on itself has only one chart, hence c' is the inclusion.
+        -- Then, compatibility of `G` *almost* yields our claim --- except that `e` is a chart
+        -- on `M` and `c` is one on `s`: we need to show that restricting `e` to `s` and composing
+        -- with `c'` yields a chart in the maximal atlas of `s`.
+        apply G.compatible_of_mem_maximalAtlas (G.subset_maximalAtlas hc)
+          (G.restriction_chart he (nmem_singleton_empty.mp h) c' hc')
```

Comment:
Turns out the `show` is not needed to compile, but feel free to keep it if it's conceptually clearer.
```suggestion
        -- The atlas of H on itself has only one chart, hence c' is the inclusion.
        -- Then, compatibility of `G` *almost* yields our claim --- except that `e` is a chart
        -- on `M` and `c` is one on `s`: we need to show that restricting `e` to `s` and composing
        -- with `c'` yields a chart in the maximal atlas of `s`.
        exact G.compatible_of_mem_maximalAtlas (G.subset_maximalAtlas hc)
          (G.restriction_chart he (nmem_singleton_empty.mp h) c' hc')
```